asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+        gdf = gdf.to_crs(epsg=27700)  # Convert to British National Grid
+        return gdf
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def generate_enw_df() -> gpd.GeoDataFrame:
+    """
+    Generate a GeoDataFrame for Electricity North West (ENW) substations.
+
+    Returns:
+        GeoDataFrame with ENW substation data.
+    """
+    enw_demand = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/dfes-2023-primary-data0.parquet",
+    )
+    enw_substations = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/ndp-pry-bsp-headroom.parquet",
+    )
+    enw_shape = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/grid_operators/ENW/ndp-pry-voronoi.parquet",
+    )
+
+    # Filter data for primary substations, current year, and best view scenario
+    enw_substations = enw_substations[
+        (enw_substations["substation_type"] == "PRIMARY")
+        & (enw_substations["year"] == "2024")
+        & (enw_substations["scenario"] == "1 - BEST VIEW")
+        & (enw_substations["status"] == "FIRM")
+    ]
+    enw_demand = enw_demand[
+        (enw_demand.group == "PRIMARY")
+        & (enw_demand.year == "2024")
+        & (enw_demand.scenario == "Best View")
+    ]
+
+    # Merge substation and demand data
+    enw = pd.merge(
+        enw_substations[["substation", "headroom_mva", "geopoint"]],
+        enw_demand[["number", "maximum_demand_mva_per_primary_substation"]],
+        left_on="substation",
+        right_on="number",
+    )
+
+    # Calculate firm capacity and add operator information
+    enw["firm_capacity_mva"] = (
+        enw["maximum_demand_mva_per_primary_substation"] + enw["headroom_mva"]
+    )
+    enw["operator"] = "ENW"
+    enw["geopoint"] = enw["geopoint"].apply(_parse_binary_geometry)
+
+    # Create GeoDataFrame and perform spatial join with shape data
+    enw_gdf = gpd.GeoDataFrame(enw, geometry="geopoint", crs=CRS)
+    enw_shape["geo_shape"] = enw_shape["geo_shape"].apply(_parse_binary_geometry)
+    enw_shape["geo_point_2d"] = enw_shape["geo_point_2d"].apply(_parse_binary_geometry)
+    enw_shape_gdf = gpd.GeoDataFrame(enw_shape, geometry="geo_shape", crs=CRS)
+    enw_df = gpd.sjoin(enw_gdf, enw_shape_gdf, how="right", predicate="within")
+    enw_df = enw_df.dropna(subset=["index_left"])
+
+    # Prepare final dataframe
+    enw_df["substation_id"] = enw_df["pry_group"] + "_" + enw_df["substation"]
+    return enw_df[
+        [
+            "substation_id",
+            "firm_capacity_mva",
+            "maximum_demand_mva_per_primary_substation",
+            "geo_shape",
+            "operator",
+        ]
+    ].rename(
+        columns={
+            "substation_id": "id",
+            "maximum_demand_mva_per_primary_substation": "peak_demand_mva",
+            "geo_shape": "geo_shape",
+        }
+    )
+
+
+def generate_npg_df() -> gpd.GeoDataFrame:
```

Comment:
```suggestion
def generate_npg_gdf() -> gpd.GeoDataFrame:
```
These should all be gdf i think